File: src/main/go/leetcode/lists/offer/Offer55_2.go
package offer

/**
 * @author wujingxinit@outlook.com
 * @date 2023/1/3 19:55
 */

func isBalanced(root *TreeNode) bool {
	if root == nil {
		return true
	}
	var hi func(root *TreeNode) int

	hi = func(root *TreeNode) int {
		if root == nil {
			return 0
		}
		return Max(hi(root.Left), hi(root.Right)) + 1
	}

	return Abs(hi(root.Left)-hi(root.Right)) <= 1 && isBalanced(root.Left) && isBalanced(root.Right)
}

File: java/src/leetcode/lists/offer/Offer55_2.java
package leetcode.lists.offer;

import common.TreeNode;

/**
 * @author jingxinwu
 * @date 2021-11-25 12:36 上午
 */
public class Offer55_2 {

    public boolean isBalanced(TreeNode root) {
        if (root == null) {
            return true;
        }
        return Math.abs(height(root.left) - height(root.right)) <= 1 && isBalanced(root.left) && isBalanced(root.right);
    }

    public int height(TreeNode root) {
        if (root == null) {
            return 0;
        }
        return Math.max(height(root.left), height(root.right)) + 1;
    }
}
